sqlx-postgres/src/types/cube.rs

```diff
@@ -0,0 +1,393 @@
+use crate::decode::Decode;
+use crate::encode::{Encode, IsNull};
+use crate::types::Type;
+use crate::{PgArgumentBuffer, PgTypeInfo, PgValueFormat, PgValueRef, Postgres};
+use sqlx_core::Error;
+
+const BYTE_WIDTH: usize = 8;
+const CUBE_TYPE_ZERO_VOLUME: usize = 128;
+const CUBE_TYPE_DEFAULT: usize = 0;
+const CUBE_DIMENSION_ONE: usize = 1;
+const DIMENSIONALITY_POSITION: usize = 3;
+const START_INDEX: usize = 4;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum PgCube {
+    Point(f64),
+    ZeroVolume(Vec<f64>),
+    OneDimensionInterval(f64, f64),
+    MultiDimension(Vec<Vec<f64>>),
+}
+
+impl Type<Postgres> for PgCube {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("cube")
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for PgCube {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        match value.format() {
+            PgValueFormat::Text => Ok(PgCube::try_from(value.as_str()?)?),
+            PgValueFormat::Binary => Ok(PgCube::try_from(value.as_bytes()?)?),
+        }
+    }
+}
+
+impl<'q> Encode<'q, Postgres> for PgCube {
+    fn produces(&self) -> Option<PgTypeInfo> {
+        Some(PgTypeInfo::with_name("cube"))
+    }
+
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        buf.extend_from_slice(&self.serialize());
+
+        IsNull::No
+    }
+}
+
+impl TryFrom<&str> for PgCube {
```

Comment:
This should be a [`FromStr`](https://doc.rust-lang.org/stable/std/str/trait.FromStr.html) impl.

It's fine to have both, but it's surprising to have a `TryFrom<&str>` impl and not a `FromStr` impl. 

The only reason to do that is if `Self` wants to borrow from the string (think zero-copy parsing) which the `FromStr` trait doesn't support due to its lack of a lifetime bound.